subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/ResourceFilterMatcher.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2011 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.plugins.ide.eclipse.model;
+
+import com.google.common.base.Objects;
+import com.google.common.collect.Sets;
+import groovy.lang.Closure;
+import groovy.lang.DelegatesTo;
+import org.gradle.api.Action;
+import org.gradle.api.InvalidUserDataException;
+import org.gradle.api.Nullable;
+import org.gradle.api.internal.ClosureBackedAction;
+
+import java.util.Set;
+
+/**
+ * The model of an Eclipse resource filter matcher.
+ * <p>
+ * The matcher decides when the containing filter (or containing matcher) applies.  The
+ * matcher configures things like whether this ResourceFilter matches resources by
+ * name, project relative path, location, last modified, etc.  Eclipse has many types
+ * of built-in matchers and it is possible to specify the id and arguments for custom
+ * matchers using this model.
+ * <p>
+ * A matcher must have an id.  It may have either a custom string argument or a set of nested
+ * child matchers (e.g. an 'or' matcher will have several nested condition matchers).
+ * <p>
+ * For more documentation on usage with examples, see {@link ResourceFilter}.
+ *
+ * @since 3.5
+ */
+public final class ResourceFilterMatcher {
+    private String id;
+    private String arguments;
+    private Set<ResourceFilterMatcher> children = Sets.newLinkedHashSet();
+
+    public ResourceFilterMatcher() {
```

Comment:
The default constructor was required to initialize an empty object that is configured by the Action while the full constructor is useful to construct instances for tests (ProjectTest, ResourceFilterTest, ResourceFilterMatcherTest).  Please let me know if you have a more idiomatic way of constructing the DSL objects for Actions, etc.